File: sdk/python/intentprotocol/crypto.py
"""Ed25519 cryptography for Intent Protocol message signing."""

from __future__ import annotations

import json
from nacl.signing import SigningKey, VerifyKey
from nacl.encoding import Base64Encoder


def generate_keypair() -> tuple[bytes, bytes]:
    """Generate an Ed25519 signing keypair.

    Returns:
        Tuple of (public_key_bytes, secret_key_bytes)
    """
    sk = SigningKey.generate()
    return bytes(sk.verify_key), bytes(sk)


def sign(message: dict, secret_key: bytes) -> str:
    """Sign a message dict with Ed25519.

    Args:
        message: Message dict to sign (will be JSON-serialized)
        secret_key: 64-byte Ed25519 secret key

    Returns:
        Signature string in "ed25519:<base64>" format
    """
    sk = SigningKey(secret_key[:32])  # NaCl seed is first 32 bytes
    msg_bytes = json.dumps(message, separators=(",", ":"), ensure_ascii=False).encode()
    signed = sk.sign(msg_bytes)
    sig_b64 = Base64Encoder.encode(signed.signature).decode()
    return f"ed25519:{sig_b64}"


def verify(message: dict, sig_str: str, public_key: bytes) -> bool:
    """Verify an Ed25519 signature on a message.

    Args:
        message: Original message dict
        sig_str: Signature in "ed25519:<base64>" format
        public_key: 32-byte Ed25519 public key

    Returns:
        True if valid
    """
    if not sig_str.startswith("ed25519:"):
        return False
    try:
        sig_bytes = Base64Encoder.decode(sig_str[8:].encode())
        vk = VerifyKey(public_key)
        msg_bytes = json.dumps(message, separators=(",", ":"), ensure_ascii=False).encode()
        vk.verify(msg_bytes, sig_bytes)
        return True
    except Exception:
        return False

File: sdk/js/src/crypto.js
import nacl from 'tweetnacl';
import tnaclUtil from 'tweetnacl-util';

const { encodeBase64, decodeBase64 } = tnaclUtil;

/**
 * Generate an Ed25519 keypair.
 * @returns {{ publicKey: string, secretKey: string }}
 */
export function generateKeypair() {
  const kp = nacl.sign.keyPair();
  return {
    publicKey: encodeBase64(kp.publicKey),
    secretKey: encodeBase64(kp.secretKey),
  };
}

/**
 * Sign a message object with an Ed25519 secret key.
 * @param {Object} message - Message to sign (will be JSON-stringified)
 * @param {string} secretKeyB64 - Base64-encoded secret key
 * @returns {string} Signature in "ed25519:<base64>" format
 */
export function sign(message, secretKeyB64) {
  const secretKey = decodeBase64(secretKeyB64);
  const msgBytes = new TextEncoder().encode(JSON.stringify(message));
  const sig = nacl.sign.detached(msgBytes, secretKey);
  return 'ed25519:' + encodeBase64(sig);
}

/**
 * Verify an Ed25519 signature on a message.
 * @param {Object} message - Original message object
 * @param {string} sigStr - Signature in "ed25519:<base64>" format
 * @param {string} publicKeyB64 - Base64-encoded public key
 * @returns {boolean}
 */
export function verify(message, sigStr, publicKeyB64) {
  if (!sigStr.startsWith('ed25519:')) return false;
  const sig = decodeBase64(sigStr.slice(8));
  const pubKey = decodeBase64(publicKeyB64);
  const msgBytes = new TextEncoder().encode(JSON.stringify(message));
  return nacl.sign.detached.verify(msgBytes, sig, pubKey);
}
